modules/connectivity/ParticipantConnectionStatus.js

```diff
@@ -380,18 +424,53 @@ export default class ParticipantConnectionStatus {
             isConnActiveByJvb = true;
         }
 
-        const isConnectionActive
-            = isConnActiveByJvb
-                && (isVideoMuted || (isInLastN && !isVideoTrackFrozen));
+        let newState = ParticipantConnectionStatus.INACTIVE;
+
+        if (isConnActiveByJvb) {
+            if (isInLastN) {
+                if (isVideoTrackFrozen) {
+                    // if participant was inactive, and is in lastN,
+                    // it is restoring
+                    const isInactive = participant.getConnectionStatus()
+                        === ParticipantConnectionStatus.INACTIVE;
+
+                    newState = isInactive
```

Comment:
Previously the method was refactored to calculate output status for given "input" values when called at any point of time. Like a function of arguments. Right now this introduces state machine where it depends on the previous state (unless I'm reading it wrong). We want to avoid that, because it makes things much more complex.

With this code if 'figureOutConnectionStatus' is called soon after it enters RESTORING it will go to INTERRUPTED, but we don't know where is this update coming from (it could be P2P status change).

I'd say this should be changed to store the time when user entered last N and if the timestamp is too old (it's taking too long to go to active) then it means it's interrupted, but before that if the video is frozen it means it's RESTORING. This logic could be encapsulated into some utility method "isParticipantRestoring" or whatever you find suitable.

Btw. probably lastN makes no sense when in P2P, so that should be taken into account as well.